core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -86,33 +86,32 @@ protected CachingCatalog(Catalog catalog, boolean caseSensitive, long expiration
   }
 
   /**
-   * CacheWriter class for removing metadata tables when their associated data table is expired
+   * RemovalListener class for removing metadata tables when their associated data table is expired
    * via cache expiration.
+   *
+   * @see com.github.benmanes.caffeine.cache.RemovalListener
    */
-  class MetadataTableInvalidatingCacheWriter implements CacheWriter<TableIdentifier, Table> {
+  class MetadataTableInvalidatingRemovalListener implements RemovalListener<TableIdentifier, Table> {
     @Override
-    public void write(TableIdentifier tableIdentifier, Table table) {
-    }
-
-    @Override
-    public void delete(TableIdentifier tableIdentifier, Table table, RemovalCause cause) {
+    public void onRemoval(TableIdentifier tableIdentifier, Table table, @NonNull RemovalCause cause) {
+      identLoggingRemovalListener.onRemoval(tableIdentifier, table, cause);
       if (RemovalCause.EXPIRED.equals(cause)) {
-        if (!MetadataTableUtils.hasMetadataTableName(tableIdentifier)) {
+        if (tableIdentifier != null && !MetadataTableUtils.hasMetadataTableName(tableIdentifier)) {
```

Comment:
Why would `tableIdentifier` be null?

Reply:
That's when the entry is GC'ed (according to caffeine javadocs:
![image](https://user-images.githubusercontent.com/5193063/147292986-d305cc86-7c9e-4c3e-a948-d39bc9b132eb.png)

Since we're dealing only with the `EXPIRED` cause, it's safe. I'll remove this check

Reply:
done